File: android-shared-lib/src/test/java/com/skedgo/android/common/util/TripSegmentListResolverTest.java
package com.skedgo.android.common.util;

import android.content.res.Resources;

import com.skedgo.android.common.BuildConfig;
import com.skedgo.android.common.TestRunner;
import com.skedgo.android.common.model.Location;
import com.skedgo.android.common.model.SegmentType;
import com.skedgo.android.common.model.Trip;
import com.skedgo.android.common.model.TripSegment;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import java.util.ArrayList;
import java.util.Arrays;

import static com.skedgo.android.common.model.TripSegment.VISIBILITY_IN_DETAILS;
import static junit.framework.Assert.assertFalse;
import static org.assertj.core.api.Assertions.assertThat;

@RunWith(TestRunner.class)
@Config(constants = BuildConfig.class)
public class TripSegmentListResolverTest {
  private Resources resources;

  @Before public void before() {
    resources = RuntimeEnvironment.application.getResources();
  }

  @Test public void shouldOnlyShowDepartureSegmentInDetails() {
    TripSegment firstSegment = new TripSegment();

    TripSegmentListResolver resolver = new TripSegmentListResolver(resources);
    TripSegment departureSegment = resolver.createDepartureSegment(firstSegment);

    assertThat(departureSegment.getVisibility()).isEqualTo(VISIBILITY_IN_DETAILS);
  }

  @Test public void shouldHaveNoTimeDifferenceForArrivalSegment() {
    TripSegment lastSegment = new TripSegment();

    // (M/D/Y @ h:m:s): 8 / 1 / 2002 @ 0:0:0 UTC
    lastSegment.setStartTimeInSecs(1028160000);

    // (M/D/Y @ h:m:s): 8 / 2 / 2002 @ 0:0:0 UTC
    lastSegment.setEndTimeInSecs(1028246400);

    TripSegmentListResolver testResolver = new TripSegmentListResolver(resources);
    TripSegment arrivalSegment = testResolver.createArrivalSegment(lastSegment);

    assertThat(arrivalSegment.getType())
        .isEqualTo(SegmentType.ARRIVAL);
    assertThat(arrivalSegment.getStartTimeInSecs())
        .isEqualTo(arrivalSegment.getEndTimeInSecs())
        .isEqualTo(lastSegment.getEndTimeInSecs());
    assertFalse(arrivalSegment.isBroken());
  }

  /**
   * @see <a href="https://redmine.buzzhives.com/issues/4197">Issue 4197</a>
   * @see <a href="https://redmine.buzzhives.com/issues/4397">Issue 4397</a>
   */
  @Test public void departureTimesBindsToFirstSegmentTimes() {
    TripSegment firstSegment = new TripSegment();
    firstSegment.setAction("Take MRT");
    firstSegment.setStartTimeInSecs(2);
    firstSegment.setEndTimeInSecs(3);

    TripSegment finalSegment = new TripSegment();
    finalSegment.setAction("Ride scooter");
    finalSegment.setStartTimeInSecs(3);
    finalSegment.setEndTimeInSecs(5);

    Trip trip = new Trip();
    trip.setSegments(new ArrayList<>(Arrays.asList(
        firstSegment,
        finalSegment
    )));

    new TripSegmentListResolver(resources)
        .setOrigin(new Location())
        .setDestination(new Location())
        .setTripSegmentList(trip.getSegments())
        .resolve();

    firstSegment.setStartTimeInSecs(3);
    firstSegment.setEndTimeInSecs(4);

    TripSegment departureSegment = trip.getSegments().get(0);
    assertThat(departureSegment.getStartTimeInSecs())
        .describedAs("Departure segment's start time must be bound to first segment's start time")
        .isEqualTo(firstSegment.getStartTimeInSecs());
    assertThat(departureSegment.getEndTimeInSecs())
        .describedAs("Departure segment's end time must be bound to first segment's start time")
        .isEqualTo(firstSegment.getStartTimeInSecs());
  }

  /**
   * @see <a href="https://redmine.buzzhives.com/issues/4197">Issue 4197</a>
   * @see <a href="https://redmine.buzzhives.com/issues/4397">Issue 4397</a>
   */
  @Test public void arrivalTimesBindsToFinalSegmentTimes() {
    TripSegment firstSegment = new TripSegment();
    firstSegment.setAction("Take MRT");
    firstSegment.setStartTimeInSecs(2);
    firstSegment.setEndTimeInSecs(3);

    TripSegment finalSegment = new TripSegment();
    finalSegment.setAction("Ride scooter");
    finalSegment.setStartTimeInSecs(3);
    finalSegment.setEndTimeInSecs(5);

    Trip trip = new Trip();
    trip.setSegments(new ArrayList<>(Arrays.asList(
        firstSegment,
        finalSegment
    )));

    new TripSegmentListResolver(resources)
        .setOrigin(new Location())
        .setDestination(new Location())
        .setTripSegmentList(trip.getSegments())
        .resolve();

    finalSegment.setStartTimeInSecs(4);
    finalSegment.setEndTimeInSecs(5);

    TripSegment arrivalSegment = trip.getSegments().get(trip.getSegments().size() - 1);
    assertThat(arrivalSegment.getStartTimeInSecs())
        .describedAs("Arrival segment's start time must be bound to final segment's end time")
        .isEqualTo(finalSegment.getEndTimeInSecs());
    assertThat(arrivalSegment.getEndTimeInSecs())
        .describedAs("Arrival segment's end time must be bound to final segment's end time")
        .isEqualTo(finalSegment.getEndTimeInSecs());
  }
}

File: CommonCoreLegacy/src/test/java/com/skedgo/tripkit/common/util/TripSegmentListResolverTest.kt
package com.skedgo.tripkit.common.util

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import org.assertj.core.api.Java6Assertions.assertThat
import org.joda.time.DateTime
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import com.skedgo.tripkit.routing.SegmentType
import com.skedgo.tripkit.routing.TripSegment
import com.skedgo.tripkit.routing.Visibilities.VISIBILITY_IN_DETAILS
import com.skedgo.tripkit.routing.Visibilities.VISIBILITY_ON_MAP
import com.skedgo.tripkit.routing.toSeconds

@RunWith(RobolectricTestRunner::class)
class TripSegmentListResolverTest {
  private val resolver by lazy {
    TripSegmentListResolver(ApplicationProvider.getApplicationContext<Context>().resources)
  }

  @Test fun shouldOnlyShowDepartureSegmentInDetails() {
    val segment = TripSegment()
    val departureSegment = resolver.createDepartureSegment(segment)
    assertThat(departureSegment.visibility).isEqualTo(VISIBILITY_IN_DETAILS)
  }

  @Test fun arrivalSegmentShouldBeVisibleInMap() {
    val segment = TripSegment()
    val arrivalSegment = resolver.createArrivalSegment(segment)
    assertThat(arrivalSegment.visibility).isEqualTo(VISIBILITY_ON_MAP)
  }

  @Test fun arrivalSegmentShouldHaveCorrectType() {
    val segment = TripSegment()
    val arrivalSegment = resolver.createArrivalSegment(segment)
    assertThat(arrivalSegment.type).isEqualTo(SegmentType.ARRIVAL)
  }

  @Test fun departureSegmentShouldHaveCorrectType() {
    val segment = TripSegment()
    val departureSegment = resolver.createDepartureSegment(segment)
    assertThat(departureSegment.type).isEqualTo(SegmentType.DEPARTURE)
  }

  @Test fun shouldHaveCorrectTimeRangeForArrivalSegment() {
    val lastSegment = TripSegment()
    lastSegment.startTimeInSecs = DateTime.parse("2012-06-30T00:00").toSeconds()
    lastSegment.endTimeInSecs = DateTime.parse("2012-06-30T00:30").toSeconds()

    val arrivalSegment = resolver.createArrivalSegment(lastSegment)

    assertThat(arrivalSegment.startTimeInSecs).isEqualTo(lastSegment.endTimeInSecs)
    assertThat(arrivalSegment.endTimeInSecs).isEqualTo(lastSegment.endTimeInSecs)
  }

  @Test fun shouldHaveCorrectTimeRangeForDepartureSegment() {
    val firstSegment = TripSegment()
    firstSegment.startTimeInSecs = DateTime.parse("2012-06-30T00:00").toSeconds()
    firstSegment.endTimeInSecs = DateTime.parse("2012-06-30T00:30").toSeconds()

    val departureSegment = resolver.createDepartureSegment(firstSegment)

    assertThat(departureSegment.startTimeInSecs).isEqualTo(firstSegment.startTimeInSecs)
    assertThat(departureSegment.endTimeInSecs).isEqualTo(firstSegment.startTimeInSecs)
  }
}
